src/Middleware/RateLimiting/src/RateLimitingMiddleware.cs

```diff
@@ -0,0 +1,120 @@
+    /// </summary>
+    /// <param name="next">The <see cref="RequestDelegate"/> representing the next middleware in the pipeline.</param>
+    /// <param name="loggerFactory">The <see cref="ILoggerFactory"/> used for logging.</param>
+    /// <param name="options">The options for the middleware.</param>
+    public RateLimitingMiddleware(RequestDelegate next, ILoggerFactory loggerFactory, IOptions<RateLimitingOptions> options)
+    {
+        _next = next ?? throw new ArgumentNullException(nameof(next));
+
+        if (loggerFactory == null)
+        {
+            throw new ArgumentNullException(nameof(loggerFactory));
+        }
+
+        _logger = loggerFactory.CreateLogger<RateLimitingMiddleware>();
+        _limiter = options.Value.Limiter;
+        _onRejected = options.Value.OnRejected;
+    }
+
+    // TODO - EventSource?
+    /// <summary>
+    /// Invokes the logic of the middleware.
+    /// </summary>
+    /// <param name="context">The <see cref="HttpContext"/>.</param>
+    /// <returns>A <see cref="Task"/> that completes when the request leaves.</returns>
+    public async Task Invoke(HttpContext context)
+    {
+        var acquireLeaseTask = TryAcquireAsync(context);
+
+        // Make sure we only ever call GetResult once on the TryEnterAsync ValueTask b/c it resets.
+        bool result;
+        RateLimitLease lease;
+
+        if (acquireLeaseTask.IsCompleted)
+        {
+            lease = acquireLeaseTask.Result;
+        }
+        else
+        {
+            lease = await acquireLeaseTask;
+        }
+        result = lease.IsAcquired;
+
+        if (result)
+        {
+            try
+            {
+                await _next(context);
+            }
+            finally
+            {
+                OnCompletion(lease);
+            }
+        }
+        else
+        {
+            RateLimiterLog.RequestRejectedLimitsExceeded(_logger);
+            context.Response.StatusCode = StatusCodes.Status503ServiceUnavailable;
+            await _onRejected(context);
+        }
+    }
+
+    private ValueTask<RateLimitLease> TryAcquireAsync(HttpContext context)
+    {
+        var lease = _limiter.Acquire(context);
+        if (lease.IsAcquired)
+        {
+            return ValueTask.FromResult(lease);
+        }
+
+        var task = _limiter.WaitAsync(context);
+        if (task.IsCompletedSuccessfully)
+        {
+            return task;
+        }
+
+        return Awaited(task);
+    }
+
+    private static void OnCompletion(RateLimitLease lease)
+    {
+        if (lease != null)
+        {
+            lease.Dispose();
+        }
+    }
+
+    private static async ValueTask<RateLimitLease> Awaited(ValueTask<RateLimitLease> task)
+    {
+        return await task;
+    }
+
+    private static partial class RateLimiterLog
+    {
+        [LoggerMessage(1, LogLevel.Debug, "Rate limits exceeded, rejecting this request with a '503 server not available' error", EventName = "RequestRejectedLimitsExceeded")]
```

Comment:
Is `1` a good enough log-id?

Reply:
Yup. It only has to be unique within the context of a logger category